toml_serialization/reader.nim

```diff
@@ -100,24 +91,54 @@ proc scanInt[T](r: var TomlReader, value: var T) =
   else:
     value = T(x)
 
-proc parseEnum*(r: var TomlReader, T: type enum): T =
-  var next = nonws(r.lex, skipLf)
+proc parseStringEnum(
+    r: var TomlReader, T: type enum, next: char,
+    stringNormalizer: static[proc(s: string): string]): T =
+  eatChar
+  var s: string
   case next
   of '\"':
-    eatChar
-    var enumStr: string
-    if scanString(r.lex, enumStr, StringType.Basic):
+    if scanString(r.lex, s, StringType.Basic):
       raiseTomlErr(r.lex, errMLStringEnum)
-    r.stringEnum(result, enumStr)
   of '\'':
-    eatChar
-    var enumStr: string
-    if scanString(r.lex, enumStr, StringType.Literal):
+    if scanString(r.lex, s, StringType.Literal):
       raiseTomlErr(r.lex, errMLStringEnum)
-    r.stringEnum(result, enumStr)
+  else:
+    raiseIllegalChar(r.lex, next)
+  try:
+    result = genEnumCaseStmt(
+      T, s,
+      default = nil, ord(T.low), ord(T.high), stringNormalizer)
+  except ValueError as err:
```

Comment:
In any case, adding raises annotation on the normalizer proc may be reasonable. It's probably OK to require that the normalizer proc should be a total function (not allowed to fail and thus `raises: []`).

Then catching `ValueError` here may become unnecessary after all (depending on the answer from the very first question in this thread).

The philosophy of the Nim serialization library is that it provides opt-in building blocks for constructing more specific serialization formats. The treatment of enums here and in the JSON library is very similar because we are designing semantics that are well aligned with the design of the language itself. We are likely to have the same policy if we end up implementing other textual formats such as YAML, XML, etc. Thus, I think some of the definitions can be potentially placed in a module inside the nim-serialization repo (e.g. There could be a shared test suite, a common definition for the normalizer type, `configureTomlDeserialization` could become `Format.configureEnumDeserialization(EnumType, ...)`, etc).